crates/ruff_linter/src/rules/ruff/rules/operation_on_closed_io.rs

```diff
@@ -0,0 +1,171 @@
+use ruff_text_size::{Ranged, TextRange};
+
+/// ## What it does
+/// Checks for usages of IO operation methods of context variables
+/// outside of the original `with` statement.
+///
+/// ## Why is this bad?
+/// Such operations will raise `ValueError: I/O operation on closed file` at runtime.
+///
+/// ## Example
+///
+/// ```python
+/// with open(".txt") as f:
+///     f.read()
+///
+/// with open(".md", "w"):
+///     f.write("")
+/// ```
+///
+/// Use instead:
+///
+/// ```python
+/// with open(".txt") as f:
+///     f.read()
+///
+/// with open(".md", "w") as f:
+///     f.write("")
+/// ```
+#[derive(ViolationMetadata)]
+pub(crate) struct OperationOnClosedIO;
+
+impl Violation for OperationOnClosedIO {
+    #[derive_message_formats]
+    fn message(&self) -> String {
+        "IO operation performed on closed IO object".to_string()
+    }
+}
+
+/// RUF061
+pub(crate) fn operation_on_closed_io(
+    checker: &Checker,
+    binding: &Binding,
+) -> Option<Vec<Diagnostic>> {
+    if !matches!(&binding.kind, BindingKind::WithItemVar) {
+        return None;
+    }
+
+    let semantic = checker.semantic();
+    let with = binding.statement(semantic)?.as_with_stmt()?;
+
+    let mut diagnostics = vec![];
+
+    for reference_id in binding.references() {
+        let reference = semantic.reference(reference_id);
+
+        if reference.end() <= with.end() {
+            continue;
+        }
+
+        let Some(expression_id) = reference.expression_id() else {
+            continue;
+        };
+
+        let Some(range) = method_reference(expression_id, semantic)
+            .or_else(|| contains_check(expression_id, semantic))
+            .or_else(|| for_loop(expression_id, semantic))
+        else {
+            continue;
+        };
+
+        let diagnostic = Diagnostic::new(OperationOnClosedIO, range);
+
+        diagnostics.push(diagnostic);
+    }
+
+    Some(diagnostics)
+}
+
+/// `f.write(...)`
+fn method_reference(expression_id: NodeId, semantic: &SemanticModel) -> Option<TextRange> {
+    let mut ancestors = semantic.expressions(expression_id);
+
+    let _io_object_ref = ancestors.next()?;
+    let attribute = ancestors.next()?.as_attribute_expr()?;
+
+    if !is_io_operation_method(&attribute.attr.id) {
+        return None;
+    }
+
+    Some(attribute.range)
+}
+
+fn is_io_operation_method(name: &str) -> bool {
+    matches!(
+        name,
+        "__iter__"
+            | "__next__"
+            | "detach"
+            | "fileno"
+            | "flush"
+            | "isatty"
+            | "read"
+            | "readline"
+            | "readlines"
+            | "reconfigure"
+            | "seek"
+            | "seekable"
+            | "tell"
+            | "truncate"
+            | "writable"
+            | "write"
+            | "writelines"
+    )
+}
+
+/// `_ in f`
+fn contains_check(expression_id: NodeId, semantic: &SemanticModel) -> Option<TextRange> {
```

Comment:
Is it necessary to check how `file` is used or what's the motivation to not flag all usages of `f` after the with statement?